File: examples/ts/if-without-else/IfWithoutElse.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract } from 'runar-testing';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'IfWithoutElse.runar.ts'), 'utf8');

describe('IfWithoutElse', () => {
  it('compiles via TestContract.fromSource', () => {
    const c = TestContract.fromSource(source, { threshold: 10n });
    expect(c.state.threshold).toBe(10n);
  });

  it('passes when one input is above the threshold', () => {
    const c = TestContract.fromSource(source, { threshold: 10n });
    const r = c.call('check', { a: 15n, b: 5n });
    expect(r.success).toBe(true);
  });

  it('passes when both inputs are above the threshold', () => {
    const c = TestContract.fromSource(source, { threshold: 10n });
    const r = c.call('check', { a: 15n, b: 20n });
    expect(r.success).toBe(true);
  });

  it('rejects when neither input is above the threshold', () => {
    const c = TestContract.fromSource(source, { threshold: 10n });
    const r = c.call('check', { a: 1n, b: 2n });
    expect(r.success).toBe(false);
  });
});

File: examples/zig/if-without-else/IfWithoutElse_test.zig
const std = @import("std");
const root = @import("../examples_test.zig");
const IfWithoutElse = @import("IfWithoutElse.runar.zig").IfWithoutElse;

fn contractPath(comptime basename: []const u8) []const u8 {
    return "if-without-else/" ++ basename;
}

fn runCompileChecks(comptime basename: []const u8) !void {
    try root.runar.compileCheckSource(std.testing.allocator, @embedFile(basename), basename);
    try root.runar.compileCheckFile(std.testing.allocator, contractPath(basename));
}

test "compile-check IfWithoutElse.runar.zig" {
    try runCompileChecks("IfWithoutElse.runar.zig");
}

test "IfWithoutElse.check passes when one input is above the threshold" {
    const c = IfWithoutElse.init(10);
    c.check(15, 5);
}

test "IfWithoutElse.check passes when both inputs are above the threshold" {
    const c = IfWithoutElse.init(10);
    c.check(15, 20);
}
